pkg/sentry/fsimpl/erofs/erofs.go

```diff
@@ -0,0 +1,362 @@
+import (
+	"os"
+	"strconv"
+	"sync"
+
+	"gvisor.dev/gvisor/pkg/abi/linux"
+	"gvisor.dev/gvisor/pkg/context"
+	"gvisor.dev/gvisor/pkg/erofs"
+	"gvisor.dev/gvisor/pkg/errors/linuxerr"
+	"gvisor.dev/gvisor/pkg/sentry/fsimpl/lock"
+	"gvisor.dev/gvisor/pkg/sentry/kernel/auth"
+	"gvisor.dev/gvisor/pkg/sentry/vfs"
+)
+
+const Name = "erofs"
+
+// Mount option names for EROFS.
+const (
+	moptImageFD = "ifd"
+)
+
+// FilesystemType implements vfs.FilesystemType.
+//
+// +stateify savable
+type FilesystemType struct{}
+
+// filesystem implements vfs.FilesystemImpl.
+//
+// +stateify savable
+type filesystem struct {
+	vfsfs vfs.Filesystem
+
+	// Immutable options.
+	mopts string
+
+	// devMinor is the filesystem's minor device number. devMinor is immutable.
+	devMinor uint32
+
+	// root is the root dentry. root is immutable.
+	root *dentry
+
+	// image is the EROFS image. image is immutable.
+	image *erofs.Image
+}
+
+// Name implements vfs.FilesystemType.Name.
+func (FilesystemType) Name() string {
+	return Name
+}
+
+// Release implements vfs.FilesystemType.Release.
+func (FilesystemType) Release(ctx context.Context) {}
+
+// GetFilesystem implements vfs.FilesystemType.GetFilesystem.
+func (fstype FilesystemType) GetFilesystem(ctx context.Context, vfsObj *vfs.VirtualFilesystem, creds *auth.Credentials, source string, opts vfs.GetFilesystemOptions) (*vfs.Filesystem, *vfs.Dentry, error) {
+	mopts := vfs.GenericParseMountOptions(opts.Data)
+
+	fd, err := getFDFromMountOptionsMap(ctx, mopts)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	f := os.NewFile(uintptr(fd), "image file")
+	image, err := erofs.OpenImage(f)
+	if err != nil {
+		f.Close()
+		return nil, nil, err
+	}
+
+	devMinor, err := vfsObj.GetAnonBlockDevMinor()
+	if err != nil {
+		image.Close()
+		return nil, nil, err
+	}
+
+	fs := &filesystem{
+		mopts:    opts.Data,
+		image:    image,
+		devMinor: devMinor,
+	}
+	fs.vfsfs.Init(vfsObj, &fstype, fs)
+
+	root := fs.newDentry(image.Root())
+	// Increase the root's reference count to 2. One reference is returned to
+	// the caller, and the other is held by fs.
+	root.IncRef()
+	fs.root = root
+
+	return &fs.vfsfs, &root.vfsd, nil
+}
+
+func getFDFromMountOptionsMap(ctx context.Context, mopts map[string]string) (int, error) {
+	ifdstr, ok := mopts[moptImageFD]
+	if !ok {
+		ctx.Warningf("erofs.getFDFromMountOptionsMap: image FD must be specified as '%s=<file descriptor>'", moptImageFD)
+		return -1, linuxerr.EINVAL
+	}
+	delete(mopts, moptImageFD)
+
+	ifd, err := strconv.Atoi(ifdstr)
+	if err != nil {
+		ctx.Warningf("erofs.getFDFromMountOptionsMap: invalid image FD: %s=%s", moptImageFD, ifdstr)
+		return -1, linuxerr.EINVAL
+	}
+
+	return ifd, nil
+}
+
+// dentry implements vfs.DentryImpl.
+//
+// +stateify savable
+type dentry struct {
+	vfsd vfs.Dentry
+
+	dentryRefs
+
+	// fs is the owning filesystem.
+	fs *filesystem
+
+	// parent is this dentry's parent directory. If this dentry is
+	// a file system root, parent is nil.
+	parent *dentry
```

Comment:
Since the dentry uses its parent, it should take a ref on the parent, so the parent is not deleted while the child is present.